File: App/Handwriter/handwriter.cpp
#include "handwriter.h"

#include <QtGui>
#include <QtWidgets>
#include <iostream>

Handwriter::Handwriter(QWidget *parent) : QWidget(parent) {
    image = new QImage(QSize(280, 280), QImage::Format_RGB32);
    clearImage();
    activate = false;
}

bool Handwriter::isModified() {
    return modified;
}

void Handwriter::clearImage() {
    image->fill(Qt::black);
    update();
    modified = false;
}

void Handwriter::saveImage() {
    if (modified) {
        QImage resizedImage = image->scaled(28, 28, Qt::KeepAspectRatio);
        QTransform transform;
        transform.rotate(90);
        resizedImage = resizedImage.transformed(transform);
        resizedImage.mirrored(true, false).save(QString("query.pgm"), "PGM");
    }
}

void Handwriter::mousePressEvent(QMouseEvent *event) {
    if (event->button() == Qt::LeftButton) {
        lastPoint = event->pos();
        activate = true;
    }
}

void Handwriter::mouseMoveEvent(QMouseEvent *event) {
    if (activate) {
        drawLineTo(event->pos());
    }
}

void Handwriter::mouseReleaseEvent(QMouseEvent *event) {
    if (event->button() == Qt::LeftButton && activate) {
        drawLineTo(event->pos());
        activate = false;
    }
}

void Handwriter::paintEvent(QPaintEvent *event) {
    QPainter painter(this);
    QRect rect = this->rect();
    painter.drawImage(rect, *image, rect);
}

void Handwriter::drawLineTo(const QPoint& endPoint) {
    QPainter painter(image);
    painter.setPen(QPen(Qt::white, 30, Qt::SolidLine, Qt::RoundCap, Qt::RoundJoin));
    painter.drawLine(lastPoint, endPoint);
    modified = true;
    update();
    lastPoint = endPoint;
}

Handwriter::~Handwriter() {
    delete image;
}

File: Python/Handwriter.py
# This Python file uses the following encoding: utf-8
from PyQt5 import QtCore
from PyQt5 import QtWidgets
from PyQt5 import QtGui

from PyQt5.QtCore import Qt, QSize
from PyQt5.QtWidgets import QWidget
from PyQt5.QtGui import QImage, QTransform, QPainter, QPen


class Handwriter(QWidget):
    def __init__(self):
        QWidget.__init__(self)
        self.image = QImage(QSize(280, 280), QImage.Format_RGB32)
        self.clearImage()
        self.activate = False

    def isModified(self):
        return self.modified

    def clearImage(self):
        self.image.fill(Qt.black)
        self.update()
        self.modified = False

    def saveImage(self):
        if self.modified:
            resizedImage = self.image.scaled(28, 28, Qt.KeepAspectRatio)
            resizedImage.save("query.pgm", "PGM")

    def mousePressEvent(self, event):
        if event.button() == Qt.LeftButton:
            self.lastPoint = event.pos()
            self.activate = True

    def mouseMoveEvent(self, event):
        if self.activate:
            self.drawLineTo(event.pos())

    def mouseReleaseEvent(self, event):
        if event.button() == Qt.LeftButton and self.activate:
            self.drawLineTo(event.pos())
            self.activate = False

    def paintEvent(self, event):
        painter = QPainter(self)
        rect = self.rect()
        painter.drawImage(rect, self.image, rect)

    def drawLineTo(self, endPoint):
        painter = QPainter(self.image)
        painter.setPen(QPen(Qt.white, 30, Qt.SolidLine, Qt.RoundCap, Qt.RoundJoin))
        painter.drawLine(self.lastPoint, endPoint)
        self.modified = True
        self.update()
        self.lastPoint = endPoint

    def __del__(self):
        del self.image
